### PR TITLE
Random Network Distillation for Torch

### PR DESCRIPTION
**Pull request by vincentpierre**
### Proposed change(s)

Adding [Random Network Distillation](https://arxiv.org/pdf/1810.12894.pdf) to the Torch trainers

Screen Shot 2020-09-11 at 9 38 35 AM

Tested with Pyramids (It get to 1.7 reward consistently)

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [x] Added tests that prove my fix is effective or that my feature works
- [x] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [x] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        self.optimizer = torch.optim.Adam(
+            self._training_network.parameters(), lr=settings.learning_rate
+        )
+        self._has_updated_once = False
+
+    def evaluate(self, mini_batch: AgentBuffer) -> np.ndarray:
+        with torch.no_grad():
+            target = self._random_network(mini_batch)
+            prediction = self._training_network(mini_batch)
+            unnormalized_rewards = torch.sum((prediction - target) ** 2, dim=1)
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py:38]: Just curious - do you know if `(prediction - target) ** 2` gets treated as `x*x` or `pow(x, 2)`? I'd imagine the former is more efficient for evaluating (both it and the gradient).

File: docs/Training-Configuration-File.md [hunk 2]
```diff
@@ -118,6 +119,18 @@ settings:
 | `gail -> use_actions`   | (default = `false`) Determines whether the discriminator should discriminate based on both observations and actions, or just observations. Set to True if you want the agent to mimic the actions from the demonstrations, and False if you'd rather have the agent visit the same states as in the demonstrations but with possibly different actions. Setting to False is more likely to be stable, especially with imperfect demonstrations, but may learn slower. |
 | `gail -> use_vail`      | (default = `false`) Enables a variational bottleneck within the GAIL discriminator. This forces the discriminator to learn a more general representation and reduces its tendency to be "too good" at discriminating, making learning more stable. However, it does increase training time. Enable this if you notice your imitation learning is unstable, or unable to learn the task at hand.                                                                       |
 
+### RND Intrinsic Reward
+
+Random Network Distillation (RND) is only available for the PyTorch trainers.
```
Comment (chriselion) [inline docs/Training-Configuration-File.md:124]: Can you explain in more detail why a user would want to enable this? What scenarios does it help solve better, and what are the drawbacks/weaknesses?

Comment (chriselion) [inline docs/Training-Configuration-File.md:124]: What happens if you try to enable this with tensorflow? Or without torch installed?

**Comment by ervteng** [comment]
Out of curiosity (no pun intended) have you tried this with Pyramids-SAC?

File: ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py (hunk 1 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py:38]: Torch converts x ** 2 to pow(x, 2). The two are equivalent, so we can pick the one we are more comfortable with.

File: docs/Training-Configuration-File.md (hunk 2 above)
Comment (vincentpierre) [reply docs/Training-Configuration-File.md:124]: You get an error that says the intrinsic reward signal type is not recognized.

**Comment by vincentpierre** [comment]
> Out of curiosity (no pun intended) have you tried this with Pyramids-SAC?

It does not do as well (It behaves a bit like curiosity, which is also having a bad time)
Will do more tests for SAC

**Comment by vincentpierre** [comment]
Some comparison between RND in orange and Curiosity in blue on a *single* cloud training run (using PPO). SAC still does not want to hear anything related to either Curiosity or RND.
Curiosity (blue) vs RND (orange)

File: ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+
+class RNDRewardProvider(BaseRewardProvider):
+    """
+    Implementation of Random Network Distillation : https://arxiv.org/pdf/1810.12894.pdf
+    """
+
+    def __init__(self, specs: BehaviorSpec, settings: RNDSettings) -> None:
+        super().__init__(specs, settings)
+        self._ignore_done = True
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py:24]: What is this for?

File: ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        self._random_network = RNDNetwork(specs, settings)
+        self._training_network = RNDNetwork(specs, settings)
+        self.optimizer = torch.optim.Adam(
+            self._training_network.parameters(), lr=settings.learning_rate
+        )
+        self._has_updated_once = False
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py:30]: I see we use this flag to zero out the reward before the training network has been trained to predict the random network. Is the idea that this initial reward is misleading in some way? If that's the case, I am not completely sold that that's true.

**Comment by andrewcoh** [comment]
Do we not mind leaving rewards unnormalized (in the paper they use the STD of full returns as a normalizer)?

File: ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py (hunk 4 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py:30]: You are right. We do the same thing in Curiosity. If there has been no update yet, the reward is set to 0. This is because if there has been no update yet, the reward is just the mse of two random networks. The first reward will just reward any action, which in my opinion is the same as rewarding none. But I see your point, I can remove this part. I do not think it will have any impact on training.

**Comment by vincentpierre** [comment]
> Do we not mind leaving rewards unnormalized (in the paper they use the STD of full returns as a normalizer)?

Yes, you are right. I took some liberties with the original paper. For example, there is also a terminal reward for end states that I simply removed and I do not consider the Agent being done zeros the future rewards.
The argument behind it is that our current Reward Provider abstraction does not allow for either of these. I think we could implement these features, but it will a rather big refactor. It still trains the Pyramids environment correctly.

File: ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py (hunk 4 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py:30]: If a user runs with `--resume`, the first reward would be ignored?

Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py:30]: Yes, you are right. I see the issue. I will remove the _has_updated_once variable

**Review by andrewcoh — APPROVED** [review]
:shipit:
